# FipeX Data Exploration

Initial exploration and data profiling of the FipeX dataset containing Brazilian FIPE vehicle price data.

## Objectives

- Understand the dataset structure and schema
- Identify missing values and duplicates
- Analyze column cardinality
- Validate the expected grain of the dataset
- Identify potential data quality issues
- Define assumptions for future transformation and modeling

## 1. Setup

In [2]:
from pathlib import Path

import pandas as pd

In [5]:
PROJECT_ROOT = Path.cwd().parent

DATA_PATH = (
    PROJECT_ROOT
    / "data"
    / "raw"
    / "fipex_prices_2026_09.parquet"
)

DATA_PATH

WindowsPath('e:/VSCODE Files/Projects/01_automotive_market_data_analysis/data/raw/fipex_prices_2026_09.parquet')

## 2. Data Loading

In [6]:
df = pd.read_parquet(DATA_PATH)

## 3. Initial Inspection

In [13]:
df.sample(25)

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia
43689,caminhão,515005-1,12-140 2p (diesel),VOLKSWAGEN,Diesel,d,1992.0,False,3036800,"R$ 30.368,00",9,2026
25619,caminhão,508038-0,VOLARE ESCOLAR V8L LONGO (diesel)(E5),MARCOPOLO,Diesel,d,2014.0,False,15350900,"R$ 153.509,00",9,2026
3798,carro,009144-8,X6 M 4.4 4x4 V8 32V Bi-Turbo Aut.,BMW,Gasolina,g,2014.0,False,16426900,"R$ 164.269,00",9,2026
29317,carro,090002-8,570S Spider 3.8 Bi-Turbo V8,Mclaren,Gasolina,g,2019.0,False,211269200,"R$ 2.112.692,00",9,2026
33650,moto,884024-5,RZR R 4 ULTIMATE Turbo 4x4 (UTV),POLARIS,Gasolina,g,2023.0,False,17581100,"R$ 175.811,00",9,2026
39020,caminhão,513254-1,R-400 A 4x2 HIGHLINE 2p (diesel)(E5),SCANIA,Diesel,d,2014.0,False,28370100,"R$ 283.701,00",9,2026
19531,carro,015001-0,Accent GLS 1.5 12V/16V Mec.,Hyundai,Gasolina,g,1996.0,False,549600,"R$ 5.496,00",9,2026
306,moto,801003-0,CITY 90,AGRALE,Gasolina,g,1998.0,False,174700,"R$ 1.747,00",9,2026
33931,carro,024088-5,206 Passion 1.6 16v 110cv 5p,Peugeot,Gasolina,g,2004.0,False,985900,"R$ 9.859,00",9,2026
4427,carro,040002-5,Buggy 1.6/ TST/ RS 1.6 4-Lug.,Baby,Gasolina,g,1990.0,False,345200,"R$ 3.452,00",9,2026


In [16]:
df.shape

(51012, 12)

In [18]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51012 entries, 0 to 51011
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   tipo_veiculo       51012 non-null  str    
 1   codigo_fipe        51012 non-null  str    
 2   nome_modelo        51012 non-null  str    
 3   nome_marca         51012 non-null  str    
 4   nome_combustivel   51012 non-null  str    
 5   sigla_combustivel  51012 non-null  str    
 6   ano_modelo         49081 non-null  float64
 7   zero_km            51012 non-null  bool   
 8   valor_centavos     51012 non-null  int64  
 9   valor_formatado    51012 non-null  str    
 10  mes_referencia     51012 non-null  int32  
 11  ano_referencia     51012 non-null  int32  
dtypes: bool(1), float64(1), int32(2), int64(1), str(7)
memory usage: 7.3 MB


## 4. Data Profiling

### 4.1. Missing Values

In [19]:
df.isna().sum()

tipo_veiculo            0
codigo_fipe             0
nome_modelo             0
nome_marca              0
nome_combustivel        0
sigla_combustivel       0
ano_modelo           1931
zero_km                 0
valor_centavos          0
valor_formatado         0
mes_referencia          0
ano_referencia          0
dtype: int64

In [20]:
null_sumary = pd.DataFrame({
    "null_count": df.isna().sum(),
    "null_percent": df.isna().mean() * 100
})

null_sumary.sort_values('null_count', ascending=False)

,null_count,null_percent
ano_modelo,1931,3.785384
tipo_veiculo,0,0.000000
nome_modelo,0,0.000000
codigo_fipe,0,0.000000
nome_marca,0,0.000000
nome_combustivel,0,0.000000
sigla_combustivel,0,0.000000
zero_km,0,0.000000
valor_centavos,0,0.000000
valor_formatado,0,0.000000


In [24]:
df[df['ano_modelo'].isna()].sample(20)

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia
23326,moto,817079-7,Z 900 RS CAFÉ,KAWASAKI,Gasolina,g,NaN,True,6869300,"R$ 68.693,00",9,2026
1707,carro,008250-3,Q8 Perf.Black 3.0 TFSI Quat. S-tr/(Híb.),Audi,Híbrido,h,NaN,True,82897300,"R$ 828.973,00",9,2026
34207,carro,024265-9,208 Style 1.0 Flex 6V 5p Mec.,Peugeot,Flex,f,NaN,True,9775900,"R$ 97.759,00",9,2026
45217,caminhão,516202-5,FH-420 4X2 2p (diesel) (E6),VOLVO,Diesel,d,NaN,True,97633700,"R$ 976.337,00",9,2026
36790,carro,025196-8,Master 2.3 dCi Furgão 16V Diesel,Renault,Diesel,d,NaN,True,27206100,"R$ 272.061,00",9,2026
50859,moto,827129-1,YZF R-15 155 ABS,YAMAHA,Gasolina,g,NaN,True,2516500,"R$ 25.165,00",9,2026
21336,caminhão,506179-2,TECTOR 17-280 4x2 (diesel) (E6),IVECO,Diesel,d,NaN,True,54213200,"R$ 542.132,00",9,2026
15140,carro,004540-3,MONTANA LT 1.2 Turbo Flex 12V 4p Mec.,GM - Chevrolet,Flex,f,NaN,True,13938900,"R$ 139.389,00",9,2026
16919,moto,810115-9,CVO ROAD GLIDE ST FLTRXSTSE,HARLEY-DAVIDSON,Gasolina,g,NaN,True,28995000,"R$ 289.950,00",9,2026
48956,carro,005520-4,T-Cross Sense 200 TSI 1.0 Flex 5p Aut.,VW - VolksWagen,Flex,f,NaN,True,11999000,"R$ 119.990,00",9,2026


In [25]:
pd.crosstab(
    df["ano_modelo"].isna(),
    df["zero_km"]
)

zero_km,False,True
ano_modelo,,
False,49081,0
True,0,1931


In [26]:
df.loc[df["ano_modelo"].isna(), "zero_km"].value_counts(dropna=False)

zero_km
True    1931
Name: count, dtype: int64

In [27]:
df.loc[df["zero_km"], "ano_modelo"].isna().value_counts()

ano_modelo
True    1931
Name: count, dtype: int64

### Insight — `ano_modelo` and `zero_km`

In this dataset, `ano_modelo` and `zero_km` follow a deterministic relationship:

- All records where `zero_km = True` have `ano_modelo = null`.
- All records where `ano_modelo = null` correspond to `zero_km = True`.
- No non-zero-km vehicle has a null `ano_modelo`.

This indicates that the null values in `ano_modelo` are not random missing data, but instead represent an expected business rule for zero-km vehicles.

Therefore, `ano_modelo` should not be imputed without explicitly preserving the `zero_km` vehicle condition.

### 4.2 Duplicate Analysis

In [29]:
df.duplicated().sum()

np.int64(0)

In [47]:
snapshot_key = [
    'codigo_fipe',
    'ano_modelo',
    'zero_km',
    'sigla_combustivel'
]

In [48]:
df.duplicated(subset=snapshot_key).sum()

np.int64(0)

In [49]:
duplicate_snapshot_rows = (
    df[df.duplicated(subset=snapshot_key, keep=False)]
    .sort_values(snapshot_key)
)

duplicate_snapshot_rows

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia


In [50]:
df[['nome_combustivel', 'sigla_combustivel']].drop_duplicates().sort_values('nome_combustivel')

,nome_combustivel,sigla_combustivel
16,Diesel,d
470,Elétrico,l
982,Flex,f
0,Gasolina,g
9621,Gás Natural,n
988,Híbrido,h
9411,Álcool,e


In [51]:
df.groupby('nome_combustivel')['sigla_combustivel'].nunique()

nome_combustivel
Diesel         1
Elétrico       1
Flex           1
Gasolina       1
Gás Natural    1
Híbrido        1
Álcool         1
Name: sigla_combustivel, dtype: int64

### Insight — Fuel Dimension

`nome_combustivel` and `sigla_combustivel` have a **1:1 relationship**.

Each fuel name maps to exactly one fuel code, and each fuel code maps to exactly one fuel name.

Therefore, only one of these columns is required to represent the fuel dimension in the snapshot grain.

### Insight — Snapshot Grain

The validated snapshot grain is:

`codigo_fipe + ano_modelo + zero_km + sigla_combustivel`

Each row represents one FIPE reference price for a unique combination of:

- FIPE code
- Model year
- Vehicle condition
- Fuel type

within a single snapshot.

In [61]:
historical_key = [
    "codigo_fipe",
    "ano_modelo",
    "zero_km",
    "sigla_combustivel",
    "mes_referencia",
    "ano_referencia"
]

### 4.3 Cardinality and Domain Analysis

In [64]:
df.nunique(dropna=False).sort_values()

mes_referencia           1
ano_referencia           1
zero_km                  2
tipo_veiculo             3
nome_combustivel         7
sigla_combustivel        7
ano_modelo              48
nome_marca             231
nome_modelo          11391
codigo_fipe          11396
valor_formatado      44368
valor_centavos       44368
dtype: int64

In [126]:
profile = pd.DataFrame({
    'dtype': df.dtypes.astype(str),
    'null_count': df.isna().sum(),
    'null_pct': df.isna().mean().mul(100).round(3),
    'unique_count': df.nunique(dropna=True),
    'unique_pct': (
        df.nunique(dropna=True) / df.notna().sum()
    ).mul(100).round(3)
})

profile.sort_values('unique_count')

,dtype,null_count,null_pct,unique_count,unique_pct
mes_referencia,int32,0,0.000,1,0.002
ano_referencia,int32,0,0.000,1,0.002
zero_km,bool,0,0.000,2,0.004
tipo_veiculo,str,0,0.000,3,0.006
nome_combustivel,str,0,0.000,7,0.014
sigla_combustivel,str,0,0.000,7,0.014
ano_modelo,float64,1931,3.785,47,0.096
nome_marca,str,0,0.000,231,0.453
nome_modelo,str,0,0.000,11391,22.330
codigo_fipe,str,0,0.000,11396,22.340


In [132]:
def classify_cardinality(row):
    n_unique = row['unique_count']

    if n_unique == 1:
        return 'constant'
    elif n_unique <= 10:
        return 'low'
    elif n_unique <= 500:
        return 'medium'
    else:
        return 'high'

In [133]:
profile.sort_values('unique_pct')

,dtype,null_count,null_pct,unique_count,unique_pct,cardinality
mes_referencia,int32,0,0.000,1,0.002,constant
ano_referencia,int32,0,0.000,1,0.002,constant
zero_km,bool,0,0.000,2,0.004,low
tipo_veiculo,str,0,0.000,3,0.006,low
nome_combustivel,str,0,0.000,7,0.014,low
sigla_combustivel,str,0,0.000,7,0.014,low
ano_modelo,float64,1931,3.785,47,0.096,low
nome_marca,str,0,0.000,231,0.453,medium
nome_modelo,str,0,0.000,11391,22.330,high
codigo_fipe,str,0,0.000,11396,22.340,high


### Cardinality Classification

For this analysis, columns will be classified primarily according to their `unique_count`:

- **Constant:** `unique_count = 1`
- **Low cardinality:** `unique_count <= 10`
- **Medium cardinality:** `10 < unique_count <= 500`
- **High cardinality:** `unique_count > 500`

`unique_pct` will be used as a complementary metric to understand the relative cardinality of each column.

These thresholds are project-specific analytical conventions adopted for this dataset and are intended to guide the appropriate depth of domain inspection. They should not be interpreted as universal cardinality standards.

### Constant Columns

`mes_referencia` and `ano_referencia` contain a single value in the current dataset.

This is expected because the file represents a single FIPE snapshot.

These columns require no further domain inspection at this stage, but they are relevant for historical datasets, where they become part of the snapshot reference and historical key.

### 4.3.1 Constant Domain Analysis

In [169]:
df[
    [
        "mes_referencia",
        "ano_referencia"
    ]
].drop_duplicates()

,mes_referencia,ano_referencia
0,9,2026


### Insights

- `mes_referencia` and `ano_referencia` each contain a single unique value across all 51,012 records.

- This behavior is consistent with the dataset representing a single FIPE monthly snapshot rather than a historical time series.

- Therefore, these columns provide no discriminatory power within the current snapshot, but they define the temporal context of every observed price.

- Both attributes remain structurally important because they identify the FIPE reference period and become part of the historical key when multiple monthly snapshots are combined.

- No domain inconsistency is observed for the reference-period columns in the current dataset.

> **Conclusion:** `mes_referencia` and `ano_referencia` behave as snapshot-level metadata. Their constant cardinality is expected and should not be interpreted as redundant data, since both attributes are required to distinguish observations across future historical snapshots.

### 4.3.2 Low-Cardinality Domain Analysis

In [134]:
categorical_cols = [
    'tipo_veiculo',
    'nome_combustivel',
    'sigla_combustivel',
    'zero_km',
    'ano_modelo'
]

In [135]:
for col in categorical_cols:
    print(f'\n{col}')
    print(df[col].value_counts(dropna=False))


tipo_veiculo
tipo_veiculo
carro       30473
caminhão    11542
moto         8997
Name: count, dtype: int64

nome_combustivel
nome_combustivel
Gasolina       24633
Diesel         16608
Flex            7150
Híbrido         1112
Elétrico         959
Álcool           466
Gás Natural       84
Name: count, dtype: int64

sigla_combustivel
sigla_combustivel
g    24633
d    16608
f     7150
h     1112
l      959
e      466
n       84
Name: count, dtype: int64

zero_km
zero_km
False    49081
True      1931
Name: count, dtype: int64

ano_modelo
ano_modelo
NaN       1931
2015.0    1705
2012.0    1696
2014.0    1665
2016.0    1616
2018.0    1585
2013.0    1575
2011.0    1561
2019.0    1552
2023.0    1552
2017.0    1524
2020.0    1517
2025.0    1515
2026.0    1499
2024.0    1458
2010.0    1449
2022.0    1394
2009.0    1370
2021.0    1369
2008.0    1321
1999.0    1284
2001.0    1266
2000.0    1258
1998.0    1222
2007.0    1204
2006.0    1200
2005.0    1112
2002.0    1079
1995.0    1072
1997.0    1066

In [136]:
def show_domain(df, column):
    return(
        df[column]
        .value_counts(dropna=False)
        .rename_axis(column)
        .reset_index(name='count')
    )

In [137]:
show_domain(df,'tipo_veiculo')

,tipo_veiculo,count
0,carro,30473
1,caminhão,11542
2,moto,8997


In [138]:
show_domain(df, 'nome_combustivel')

,nome_combustivel,count
0,Gasolina,24633
1,Diesel,16608
2,Flex,7150
3,Híbrido,1112
4,Elétrico,959
5,Álcool,466
6,Gás Natural,84


In [139]:
show_domain(df, 'sigla_combustivel')

,sigla_combustivel,count
0,g,24633
1,d,16608
2,f,7150
3,h,1112
4,l,959
5,e,466
6,n,84


In [143]:
show_domain(df, 'ano_modelo').sort_values(by='ano_modelo', ascending=False)

,ano_modelo,count
35,2027.0,647
13,2026.0,1499
12,2025.0,1515
14,2024.0,1458
9,2023.0,1552
16,2022.0,1394
18,2021.0,1369
11,2020.0,1517
8,2019.0,1552
5,2018.0,1585


### Insight — Low-Cardinality Domains

The observed domains are consistent with the expected structure of the FIPE snapshot.

#### Vehicle Type

`tipo_veiculo` contains three observed categories:

- `carro`
- `caminhão`
- `moto`

The domain is small, well-defined, and shows no apparent formatting inconsistencies.

#### Fuel Dimension

`nome_combustivel` and `sigla_combustivel` contain seven corresponding categories, with matching frequencies across both columns.

This reinforces the previously validated **1:1 relationship** between fuel name and fuel code.

The observed mappings are:

- `Gasolina` ↔ `g`
- `Diesel` ↔ `d`
- `Flex` ↔ `f`
- `Híbrido` ↔ `h`
- `Elétrico` ↔ `l`
- `Álcool` ↔ `e`
- `Gás Natural` ↔ `n`

No unmatched or unexpected fuel categories were observed.

#### Model Year

`ano_modelo` contains a broader but still manageable domain of model years.

The observed domain includes model year `2027`, confirming that this value is a legitimate model year in the dataset.

Therefore, null values in `ano_modelo` for zero-km vehicles should not be imputed with `2027`, since this would conflate an actual model year with the distinct business meaning represented by `zero_km = True`.

### 4.3.3 Medium-Cardinality Domain Analysis

In [82]:
df['nome_marca'].value_counts().sample(50)

nome_marca
Acura                 15
Walk                  21
Mahindra              22
HAOJUE                80
Subaru               349
Benelli                9
WATTS                 14
GM - Chevrolet      2641
RIGUETE               95
GARINNI               31
FYM                   24
DAYANG                43
Citroën              863
SUNDOWN               90
Mitsubishi           973
Fyber                 34
MASCARELLO            36
MG                    29
KAWASAKI             587
SIAMOTO                1
VENTO                  5
JOHNNYPAG              6
Caoa Chery/Chery     115
YAMAHA               816
CFMOTO                 8
HITECH ELECTRIC       82
BIMOTA                19
SWM                    2
AMAZONAS              16
EFFA-JMC               6
JPX                   40
Jaguar               281
HUSABERG              13
DAYUN                 21
Hyundai              829
POLARIS              135
PEUGEOT               25
DERBI                  8
ORCA                   6
BETA          

In [83]:
df['nome_marca'].value_counts().head(50)

nome_marca
VW - VolksWagen    2722
GM - Chevrolet     2641
Fiat               2497
MERCEDES-BENZ      2328
Ford               2279
Mercedes-Benz      2261
SCANIA             2044
BMW                1915
VOLVO              1593
Renault            1430
VOLKSWAGEN         1374
Audi               1163
HONDA              1150
Toyota             1130
IVECO              1127
FORD               1055
Peugeot            1016
Mitsubishi          973
Porsche             918
Citroën             863
Hyundai             829
YAMAHA              816
Land Rover          742
MARCOPOLO           727
Nissan              697
SUZUKI              663
Volvo               632
Honda               611
KAWASAKI            587
HARLEY-DAVIDSON     586
TRIUMPH             499
Kia Motors          496
DAF                 445
Jeep                401
AGRALE              372
Subaru              349
Suzuki              343
KTM                 332
DUCATI              308
SHINERAY            293
Jaguar              281
MINI 

In [84]:
df['nome_marca'].value_counts().tail(50)

nome_marca
CBT Jipe           8
CFMOTO             8
CHANGAN            8
Caoa Changan       8
DERBI              8
Jaecoo             8
LERIVO             8
VOGE               8
VOLTZ              8
BUENO              7
HYUNDAI            7
Pontiac            7
BETA               6
EFFA-JMC           6
FOX                6
JOHNNYPAG          6
LEVA               6
Mobílli            6
ORCA               6
Plymouth           6
SHACMAN            6
ARROW              5
Cross Lander       5
Denza              5
EMME               5
GREEN              5
Miura              5
VENTO              5
CAB Motors         4
HARTFORD           4
JIAPENG VOLCANO    4
LAVRALE            4
LOBINI             4
TARGOS             4
GREAT WALL         3
MAXIBUS            3
MRX                3
Malaguti           3
Mercury            3
Rover              3
BRAVA              2
DFSK               2
Lotus              2
SERES              2
SWM                2
Saab               2
Saturn             2
LA

In [146]:
(
    df.groupby(df['nome_marca'].str.lower())['nome_marca']
    .nunique()
    .sort_values(ascending=False)
    .head(20)
)

nome_marca
agrale           2
fiat             2
honda            2
ford             2
mercedes-benz    2
hyundai          2
peugeot          2
suzuki           2
volvo            2
asia motors      1
aprilia          1
arrow            1
adly             1
acura            1
bajaj            1
bee              1
benelli          1
bepobus          1
beta             1
bimota           1
Name: nome_marca, dtype: int64

In [86]:
df[
    df['nome_marca'].str.lower() == 'volvo'
]['nome_marca'].value_counts()

nome_marca
VOLVO    1593
Volvo     632
Name: count, dtype: int64

In [147]:
brand_variants = (
    df.groupby(df['nome_marca'].str.lower())['nome_marca']
    .agg(
        variant_count='nunique',
        variants=lambda x: sorted(x.unique())
    )
    .query('variant_count > 1')
)

brand_variants

,variant_count,variants
nome_marca,,
agrale,2,"[AGRALE, Agrale]"
fiat,2,"[FIAT, Fiat]"
ford,2,"[FORD, Ford]"
honda,2,"[HONDA, Honda]"
hyundai,2,"[HYUNDAI, Hyundai]"
mercedes-benz,2,"[MERCEDES-BENZ, Mercedes-Benz]"
peugeot,2,"[PEUGEOT, Peugeot]"
suzuki,2,"[SUZUKI, Suzuki]"
volvo,2,"[VOLVO, Volvo]"


#### Case-Insensitive Brand Variants

A case-insensitive grouping of `nome_marca` identified multiple textual variants that represent the same manufacturer.

The following brands currently appear with inconsistent capitalization:

| Normalized brand | Variants |
|---|---|
| `agrale` | `AGRALE`, `Agrale` |
| `fiat` | `FIAT`, `Fiat` |
| `ford` | `FORD`, `Ford` |
| `honda` | `HONDA`, `Honda` |
| `hyundai` | `HYUNDAI`, `Hyundai` |
| `mercedes-benz` | `MERCEDES-BENZ`, `Mercedes-Benz` |
| `peugeot` | `PEUGEOT`, `Peugeot` |
| `suzuki` | `SUZUKI`, `Suzuki` |
| `volvo` | `VOLVO`, `Volvo` |

This confirms that `nome_marca` is not fully standardized with respect to letter casing.

### 4.3.4 High-Cardinality Domain Analysis

In [149]:
df['nome_modelo'].value_counts().sample(20)

nome_modelo
Sprinter 315 Furgão Str. L T.A 2.0 Dies.     4
DAILY CHASSI 40S14 2p (diesel) (E5)          3
THUNDER BOY/ WAY EXECUTIVO 1p (diesel)      12
CF FAS 290 6x2 Space Cab Aut (Die.)(E6)      5
Z3 2.8 Aut.                                  4
P-340 A 4x2 GNC 2p (GNV)(E5)                 2
can-am Outlander L MAX 450 4X4 Quadr.        2
F-Type S Supercharged Conversivel 3.0 V6     4
Grand Vitara 2.0 16V 4x2/4x4 5p Aut.        14
Lancer GLX                                   8
C3 Picasso Tendance 1.5 Flex 8V 5p Mec.      1
DAILY FURGÃO CITY 38.13 4p (diesel)          2
CANYON 500                                   5
180-D Van 2.4 Diesel                         4
C4 LOUNGE S 1.6 Turbo Flex Aut.              2
Sportage EX 2.0 16V Mec.                     5
DAKOTA WARLOCK 2.2 TB CD 4x4 Aut.            2
Palio Rua 1.0 Fire Flex 8V 5p                1
STRALIS HI-WAY 600-S48T 6x2 (diesel)(E5)    11
VOLARE ATT. 8 E. LONGO EXECUT (die.)(E6)     6
Name: count, dtype: int64

In [104]:
df['nome_modelo'].value_counts().head(20)

nome_modelo
Buggy Plus 1.6 8V                          39
Buggy/M-8/M-8 Long 1.6                     37
Buggy 1.6/ TST/ RS 1.6 4-Lug.              36
Kombi Furgão                               34
Kombi Standard/ Luxo/ Série Prata          34
GSX 1300-R HAYABUSA                        29
ROAD KING CLASSIC  FLHRC/ CUSTOM FLHRSI    25
CG 125 CARGO/ CARGO KS/125i CARGO          25
911 Turbo Coupe 3.6/3.8 (991/992)          25
Buggy IV e V                               24
A-20 Custom/ C-20 Luxe 4.1                 24
A-20 Custom/ C-20 S 4.1                    24
FAT BOY FLSTF                              24
HERITAGE SOFTAIL CLASSIC FLSTC             24
GOLD WING 1500/ 1800                       24
Forester 2.0/2.0 S/2.0 ES 4x4 Aut.         23
Buggy 2000W 1.8 8V/ 1.8 8V Flex            22
Corvette 5.7/ 6.0, 6.2 Targa/Stingray      22
NINJA ZX-10/ ZX-10R 1000cc/30TH Aniv.      22
Kombi Pick-Up                              22
Name: count, dtype: int64

In [106]:
df['nome_modelo'].value_counts().tail(20)

nome_modelo
XC 90 T-8 INSC. EXPRESS. 2.0 (Híbrido)    1
FZ25 FAZER THOR FLEX                      1
MAJESTY YP 250                            1
MT-03 321 ABS Homem de Ferro              1
MT-03 660cc                               1
NMAX Connected MANDALORIAN 160 ABS        1
NMAX HOMEM ARANHA 160 ABS                 1
T115 CRYPTON ED PENELOPE                  1
TRX 850                                   1
TT-R 125 LWE                              1
WR 426 F                                  1
XJ 600                                    1
XJR 1200                                  1
XMAX 250 ABS Darth Vader                  1
XMAX 250 CONNECTED                        1
XT 1200 Z S.TÉNÉRÉ 60th ANNIVERSARY       1
YBR 125 FACTOR PRO E                      1
YBR 125 FACTOR PRO K                      1
YZF R-1M 1000                             1
YZF R-3 World GP 60th ANNIVERSARY ED.     1
Name: count, dtype: int64

In [105]:
df['nome_modelo'].value_counts().describe()

count    11391.000000
mean         4.478272
std          3.238356
min          1.000000
25%          2.000000
50%          4.000000
75%          6.000000
max         39.000000
Name: count, dtype: float64

In [113]:
model_counts = df['nome_modelo'].value_counts()

(model_counts == 1).sum()

np.int64(1332)

In [114]:
(model_counts == 1).mean() * 100

np.float64(11.693442191203582)

In [148]:
name_variants = (
    df.groupby(df['nome_modelo'].str.lower())['nome_modelo']
    .agg(
        variant_count='nunique',
        variants=lambda x: sorted(x.unique())
    )
    .query('variant_count > 1')
)

name_variants

,variant_count,variants
nome_modelo,,


### Insight — High-Cardinality Model Domain

`nome_modelo` is a high-cardinality categorical attribute with **11,391 distinct values**.

The frequency distribution shows that model descriptions are highly granular:

- Mean frequency per model: approximately **4.48**
- Median frequency per model: **4**
- 25% of models appear at most **2 times**
- 75% of models appear at most **6 times**
- Maximum observed frequency: **39**
- **1,332 models** appear exactly once, representing approximately **11.69%** of all distinct model descriptions

This indicates that the high cardinality of `nome_modelo` is largely structural and reflects detailed vehicle configurations, trims, engines, body types, and other model-level specifications.

A case-insensitive comparison found no duplicated model names caused solely by capitalization differences.

Therefore, unlike `nome_marca`, there is no evidence at this stage that letter casing is materially inflating the cardinality of `nome_modelo`.

In [151]:
fipe_to_model = (
    df.groupby('codigo_fipe')['nome_modelo']
    .nunique()
    .sort_values(ascending=False)
)

fipe_to_model.head(20)

codigo_fipe
001001-4    1
001002-2    1
001003-0    1
001004-9    1
001005-7    1
001006-5    1
001007-3    1
001008-1    1
001009-0    1
001010-3    1
001011-1    1
001012-0    1
001013-8    1
001014-6    1
001015-4    1
001016-2    1
001017-0    1
001018-9    1
001020-0    1
001021-9    1
Name: nome_modelo, dtype: int64

In [153]:
(fipe_to_model > 1).sum()

np.int64(0)

In [154]:
model_to_fipe = (
    df.groupby('nome_modelo')['codigo_fipe']
    .nunique()
    .sort_values(ascending=False)
)

model_to_fipe.head(20)

nome_modelo
FENIX GOLD 240                                2
16-220 T 3-Eixos 2p (diesel)                  2
16-220 Turbo 2p (diesel)                      2
CR 125                                        2
CR 250                                        2
11-180 Delivery 4x4 2p (diesel)(E6)           1
up! black/white/red I MOTION 1.0 Flex 5p      1
up! black/white/red 1.0 TSI TFlex 12V 5p      1
106 Selection 1.0 3p                          1
Ë-Jumpy Cargo 136cv (Elétrico)                1
 Range R. VEL. R-Dyn. S 2.0 Aut. (Híbrido)    1
001 Flagship AWD (Elétrico)                   1
001 Premium AWD (Elétrico)                    1
10-160 E Delivery 2p (diesel)(E5)             1
10-160 E Delivery Plus 6x2 (diesel)(E5)       1
10-16DT 3.8 4x2 Diesel(E5)                    1
100 2.8 V6                                    1
100 2.8 V6 Avant                              1
100 S-4 2.2 Avant Turbo                       1
10000 / 10000 S  2p (diesel) (E5)             1
Name: codigo_fipe, dtype: in

In [155]:
(model_to_fipe > 1).sum()

np.int64(5)

In [156]:
model_fipe_variants = (
    df.groupby('nome_modelo')['codigo_fipe']
    .agg(
        fipe_count='nunique',
        fipe_codes=lambda x: sorted(x.unique())
    )
    .query('fipe_count > 1')
    .sort_values('fipe_count', ascending=False)
)

model_fipe_variants

,fipe_count,fipe_codes
nome_modelo,,
16-220 T 3-Eixos 2p (diesel),2,"[505007-3, 515043-4]"
16-220 Turbo 2p (diesel),2,"[505008-1, 515044-2]"
CR 125,2,"[811053-0, 812008-0]"
CR 250,2,"[811050-6, 812011-0]"
FENIX GOLD 240,2,"[845014-5, 872001-0]"


In [157]:
fipe_model_variants = (
    df.groupby('codigo_fipe')['nome_modelo']
    .agg(
        model_count='nunique',
        models=lambda x: sorted(x.unique())
    )
    .query('model_count > 1')
    .sort_values('model_count', ascending=False)
)

fipe_model_variants

,model_count,models
codigo_fipe,,


In [158]:
relationship_summary = pd.Series({
    'unique_models': df['nome_modelo'].nunique(),
    'unique_fipe_codes': df['codigo_fipe'].nunique(),
    'models_with_multiple_fipe_codes': (model_to_fipe > 1).sum(),
    'fipe_codes_with_multiple_models': (fipe_to_model > 1).sum()
})

relationship_summary

unique_models                      11391
unique_fipe_codes                  11396
models_with_multiple_fipe_codes        5
fipe_codes_with_multiple_models        0
dtype: int64

In [159]:
df[
    df['nome_modelo'].isin(model_fipe_variants.index)
]

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia
16634,caminhão,505007-3,16-220 T 3-Eixos 2p (diesel),GMC,Diesel,d,1996.0,False,3480100,"R$ 34.801,00",9,2026
16635,caminhão,505007-3,16-220 T 3-Eixos 2p (diesel),GMC,Diesel,d,1997.0,False,3728000,"R$ 37.280,00",9,2026
16636,caminhão,505007-3,16-220 T 3-Eixos 2p (diesel),GMC,Diesel,d,1998.0,False,3821200,"R$ 38.212,00",9,2026
16637,caminhão,505007-3,16-220 T 3-Eixos 2p (diesel),GMC,Diesel,d,1999.0,False,4543200,"R$ 45.432,00",9,2026
16638,caminhão,505007-3,16-220 T 3-Eixos 2p (diesel),GMC,Diesel,d,2000.0,False,4763300,"R$ 47.633,00",9,2026
...,...,...,...,...,...,...,...,...,...,...,...,...
44088,caminhão,515044-2,16-220 Turbo 2p (diesel),VOLKSWAGEN,Diesel,d,1996.0,False,4758400,"R$ 47.584,00",9,2026
44089,caminhão,515044-2,16-220 Turbo 2p (diesel),VOLKSWAGEN,Diesel,d,1997.0,False,5545200,"R$ 55.452,00",9,2026
44090,caminhão,515044-2,16-220 Turbo 2p (diesel),VOLKSWAGEN,Diesel,d,1998.0,False,5790000,"R$ 57.900,00",9,2026
44091,caminhão,515044-2,16-220 Turbo 2p (diesel),VOLKSWAGEN,Diesel,d,1999.0,False,6848600,"R$ 68.486,00",9,2026


In [162]:
df[
    df['nome_modelo'].isin(model_fipe_variants.index)
][
    [
        'nome_marca',
        'nome_modelo',
        'codigo_fipe',
        'ano_modelo',
        'zero_km',
        'sigla_combustivel'
    ]
].sort_values(
    ['nome_modelo', 'codigo_fipe']
)

,nome_marca,nome_modelo,codigo_fipe,ano_modelo,zero_km,sigla_combustivel
16634,GMC,16-220 T 3-Eixos 2p (diesel),505007-3,1996.0,False,d
16635,GMC,16-220 T 3-Eixos 2p (diesel),505007-3,1997.0,False,d
16636,GMC,16-220 T 3-Eixos 2p (diesel),505007-3,1998.0,False,d
16637,GMC,16-220 T 3-Eixos 2p (diesel),505007-3,1999.0,False,d
16638,GMC,16-220 T 3-Eixos 2p (diesel),505007-3,2000.0,False,d
...,...,...,...,...,...,...
18766,HUSQVARNA,CR 250,812011-0,2001.0,False,g
29001,MVK,FENIX GOLD 240,845014-5,2009.0,False,g
35828,REGAL RAPTOR,FENIX GOLD 240,872001-0,2009.0,False,g
35829,REGAL RAPTOR,FENIX GOLD 240,872001-0,2011.0,False,g


In [163]:
df[
    df['nome_modelo'].isin(model_fipe_variants.index)
][
    ['nome_marca', 'nome_modelo', 'codigo_fipe']
].drop_duplicates().sort_values(
    ['nome_modelo', 'codigo_fipe']
)

,nome_marca,nome_modelo,codigo_fipe
16634,GMC,16-220 T 3-Eixos 2p (diesel),505007-3
44073,VOLKSWAGEN,16-220 T 3-Eixos 2p (diesel),515043-4
16641,GMC,16-220 Turbo 2p (diesel),505008-1
44083,VOLKSWAGEN,16-220 Turbo 2p (diesel),515044-2
18131,HONDA,CR 125,811053-0
18762,HUSQVARNA,CR 125,812008-0
18144,HONDA,CR 250,811050-6
18765,HUSQVARNA,CR 250,812011-0
29001,MVK,FENIX GOLD 240,845014-5
35828,REGAL RAPTOR,FENIX GOLD 240,872001-0


In [164]:
brand_model_to_fipe = (
    df.groupby(['nome_marca', 'nome_modelo'])['codigo_fipe']
    .nunique()
    .sort_values(ascending=False)
)

brand_model_to_fipe.head(20)

nome_marca  nome_modelo                         
ADLY        ATV 100                                 1
            ATV 50                                  1
            JAGUAR JT 100                           1
            JAGUAR JT 50                            1
            RT 50                                   1
AGRALE      10000 / 10000 S  2p (diesel) (E5)       1
            10000 LX 2p (diesel) (E5)               1
            13000 Turbo 2p (diesel)                 1
            13000 Turbo 3-Eixos 2p (diesel)         1
            14000 / 14000 S 2p (diesel) (E5)        1
            14000 / 14000 S 6x2 2p (diesel) (E5)    1
            14000 LX 2p (diesel) (E5)               1
            1600 D-RD 2p (diesel)                   1
            1600 D-RS 2p (diesel)                   1
            1800 D-RD 2p (diesel)                   1
            1800 D-RS 2p (diesel)                   1
            4500 D-RD 2p (diesel)                   1
            4500 D-RS 2p (diesel)

In [165]:
(brand_model_to_fipe > 1).sum()

np.int64(0)

In [166]:
fipe_to_brand_model = (
    df.groupby('codigo_fipe')[['nome_marca', 'nome_modelo']]
    .nunique()
)

fipe_to_brand_model[
    (fipe_to_brand_model['nome_marca'] > 1) |
    (fipe_to_brand_model['nome_modelo'] > 1)
]

,nome_marca,nome_modelo
codigo_fipe,,


### Insight — `codigo_fipe` and `nome_modelo`

`nome_modelo` and `codigo_fipe` have very similar cardinalities:

- `nome_modelo`: **11,391** unique values
- `codigo_fipe`: **11,396** unique values

The difference is explained by a small set of model names that are shared across different manufacturers.

The following five model names are associated with more than one FIPE code:

- `16-220 T 3-Eixos 2p (diesel)`
- `16-220 Turbo 2p (diesel)`
- `CR 125`
- `CR 250`
- `FENIX GOLD 240`

In these cases, the ambiguity is resolved by `nome_marca`. For example, the same `nome_modelo` can refer to different vehicles when associated with different manufacturers.

The following relationships were validated:

- Each `codigo_fipe` maps to exactly one `nome_marca` + `nome_modelo` combination.
- Each `nome_marca` + `nome_modelo` combination maps to exactly one `codigo_fipe`.
- `nome_modelo` alone is not globally unique across manufacturers.

Therefore, `codigo_fipe` behaves as a unique business identifier for the brand-model combination, while `nome_modelo` should not be used independently as a unique vehicle identifier.

This also confirms that the small difference in cardinality between `nome_modelo` and `codigo_fipe` is expected and does not indicate a data quality issue.

In [172]:
df["valor_centavos"].describe()

count    5.101200e+04
mean     1.574897e+07
std      3.687481e+07
min      8.720000e+04
25%      2.485450e+06
50%      5.772450e+06
75%      1.512800e+07
max      9.749142e+08
Name: valor_centavos, dtype: float64

In [174]:
df[
    df["valor_centavos"] <= 0
]

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia


In [175]:
(df["valor_centavos"] <= 0).sum()

np.int64(0)

In [176]:
df[
    [
        "codigo_fipe",
        "nome_marca",
        "nome_modelo",
        "ano_modelo",
        "valor_centavos",
        "valor_formatado"
    ]
].nsmallest(
    10,
    "valor_centavos"
)

,codigo_fipe,nome_marca,nome_modelo,ano_modelo,valor_centavos,valor_formatado
41747,814001-4,TRAXX,CJ 50-F,1995.0,87200,"R$ 872,00"
40268,824005-1,SUNDOWN,PALIO 50cc,1996.0,87700,"R$ 877,00"
33463,820006-8,ORCA,QM 50,1998.0,89300,"R$ 893,00"
40223,824003-5,SUNDOWN,ERGON 50cc,1996.0,89800,"R$ 898,00"
40269,824005-1,SUNDOWN,PALIO 50cc,1997.0,90400,"R$ 904,00"
341,801012-9,AGRALE,JUNIOR 50,1995.0,90600,"R$ 906,00"
41748,814001-4,TRAXX,CJ 50-F,1996.0,91100,"R$ 911,00"
3991,804012-5,BRANDY,ZANELLA NEW FIRE 50 FULL,1996.0,91200,"R$ 912,00"
1980,837001-0,BAJAJ,CHAMPION 45 100cc,1997.0,91300,"R$ 913,00"
354,801014-5,AGRALE,SST 13.5 125cc,1990.0,92200,"R$ 922,00"


In [177]:
df[
    [
        "codigo_fipe",
        "nome_marca",
        "nome_modelo",
        "ano_modelo",
        "valor_centavos",
        "valor_formatado"
    ]
].nlargest(
    10,
    "valor_centavos"
)

,codigo_fipe,nome_marca,nome_modelo,ano_modelo,valor_centavos,valor_formatado
24249,078017-0,LAMBORGHINI,AVENTADOR LP 770-4 SVJ,2022.0,974914200,"R$ 9.749.142,00"
24248,078017-0,LAMBORGHINI,AVENTADOR LP 770-4 SVJ,2021.0,951135800,"R$ 9.511.358,00"
24247,078017-0,LAMBORGHINI,AVENTADOR LP 770-4 SVJ,2020.0,886041100,"R$ 8.860.411,00"
24234,078020-0,LAMBORGHINI,AVENTADOR 770-4 SVJ ROADSTER V12,2021.0,861832300,"R$ 8.618.323,00"
24306,078025-1,LAMBORGHINI,REVUELTO Coupe LP 1015-4 (Hib),NaN,845750000,"R$ 8.457.500,00"
8686,031059-0,Ferrari,12Cilinrdri Spider 6.5 V12 830cv,NaN,826606500,"R$ 8.266.065,00"
24305,078025-1,LAMBORGHINI,REVUELTO Coupe LP 1015-4 (Hib),2025.0,811525500,"R$ 8.115.255,00"
8803,031063-8,Ferrari,849 Testarossa Coupe 4.0 V8 830cv,NaN,810000000,"R$ 8.100.000,00"
8683,031058-1,Ferrari,12Cilinrdri Coupe 6.5 V12 830cv,NaN,790000000,"R$ 7.900.000,00"
8685,031059-0,Ferrari,12Cilinrdri Spider 6.5 V12 830cv,2026.0,784920000,"R$ 7.849.200,00"


In [178]:
centavos_to_formatado = (
    df.groupby("valor_centavos")["valor_formatado"]
    .nunique()
)

centavos_to_formatado[
    centavos_to_formatado > 1
]

Series([], Name: valor_formatado, dtype: int64)

In [179]:
formatado_to_centavos = (
    df.groupby("valor_formatado")["valor_centavos"]
    .nunique()
)

formatado_to_centavos[
    formatado_to_centavos > 1
]

Series([], Name: valor_centavos, dtype: int64)

In [180]:
df[
    [
        "valor_centavos",
        "valor_formatado"
    ]
].drop_duplicates().head(20)

,valor_centavos,valor_formatado
0,320000,"R$ 3.200,00"
1,340300,"R$ 3.403,00"
2,372700,"R$ 3.727,00"
3,235800,"R$ 2.358,00"
4,246300,"R$ 2.463,00"
5,141800,"R$ 1.418,00"
6,152200,"R$ 1.522,00"
7,165000,"R$ 1.650,00"
8,181100,"R$ 1.811,00"
9,119900,"R$ 1.199,00"


In [181]:
valor_formatado_numeric = (
    df["valor_formatado"]
    .str.replace("R$ ", "", regex=False)
    .str.replace(".", "", regex=False)
    .str.replace(",", ".", regex=False)
    .astype(float)
)

In [182]:
valor_formatado_centavos = (
    valor_formatado_numeric * 100
).round().astype("int64")

In [183]:
format_mismatch = df[
    valor_formatado_centavos
    != df["valor_centavos"]
]

format_mismatch

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia


In [184]:
format_pattern = r"^R\$ \d{1,3}(\.\d{3})*,\d{2}$"

invalid_format = df[
    ~df["valor_formatado"].str.match(
        format_pattern,
        na=False
    )
]

invalid_format

,tipo_veiculo,codigo_fipe,nome_modelo,nome_marca,nome_combustivel,sigla_combustivel,ano_modelo,zero_km,valor_centavos,valor_formatado,mes_referencia,ano_referencia


### Insights

- `valor_centavos` contains only positive values, with no zero or negative prices identified in the current snapshot.

- The observed price range goes from **R$ 872,00** to **R$ 9.749.142,00**, indicating a wide but plausible automotive market spectrum, from low-value motorcycles to high-end exotic vehicles.

- The lower and upper extremes were manually inspected and appear semantically plausible based on vehicle brand, model, and model year. Therefore, these extreme values should not be treated as data quality errors solely because they are statistical outliers.

- `valor_centavos` and `valor_formatado` present a bidirectional one-to-one relationship:
  - each `valor_centavos` maps to exactly one `valor_formatado`;
  - each `valor_formatado` maps to exactly one `valor_centavos`.

- After parsing `valor_formatado` and converting it back to integer cents, no mismatches were found when compared with `valor_centavos`.

- All values in `valor_formatado` conform to the expected Brazilian currency pattern (`R$ X.XXX,XX`).

- These validations indicate that both columns represent the same monetary information consistently, with `valor_centavos` serving as the numeric analytical representation and `valor_formatado` serving as the presentation-oriented representation.

> **Conclusion:** no monetary-domain inconsistencies were identified. `valor_centavos` should be considered the canonical field for calculations and downstream analytical processing, while `valor_formatado` is a valid derived representation intended for display.

## 4.4. Data Profiling Summary

The exploratory analysis identified structural, domain, and relational constraints that were consolidated into:

- `docs/data_dictionary.md`
- `docs/data_quality_rules.md`

These documents will serve as the formal metadata and data quality specification for the downstream pipeline.